rmw_implementation/CMakeLists.txt

```diff
@@ -15,65 +15,50 @@ find_package(ament_cmake REQUIRED)
 
 find_package(rcutils REQUIRED)
 
-# provides FindPoco.cmake and Poco on platforms without it
-find_package(poco_vendor)
-find_package(Poco COMPONENTS Foundation)
 find_package(rmw_implementation_cmake REQUIRED)
 
 if(BUILD_TESTING)
   find_package(ament_lint_auto REQUIRED)
   ament_lint_auto_find_test_dependencies()
 endif()
 
+find_package(rmw REQUIRED)
 get_default_rmw_implementation(RMW_IMPLEMENTATION)
 
-get_available_rmw_implementations(RMW_IMPLEMENTATIONS)
-message(STATUS "")
-message(STATUS "Found these rmw implementations:")
-foreach(impl IN LISTS RMW_IMPLEMENTATIONS)
-  message(STATUS "  '${impl}'")
-endforeach()
-
 message(STATUS "")
 message(STATUS "Default rmw implementation: '${RMW_IMPLEMENTATION}'")
 message(STATUS "")
 
-option(RMW_IMPLEMENTATION_FORCE_DYNAMIC_LOADING
-  "Force use of dynamic loading (via poco), even if there is only one implementation available"
-  OFF)
-
-# Nominally use dynamic loading if poco is available.
-set(RMW_IMPLEMENTATION_SUPPORTS_POCO ${Poco_FOUND})
-# Determine if we're going to actually support dynamic loading.
-if(RMW_IMPLEMENTATION_FORCE_DYNAMIC_LOADING)
-  # Force usage.
-  if(NOT Poco_FOUND)
-    message(FATAL_ERROR "Dynamic loading forced, but poco is not found!")
-  endif()
-  set(RMW_IMPLEMENTATION_SUPPORTS_POCO TRUE)
-elseif(NOT RMW_IMPLEMENTATIONS MATCHES ";")
-  set(RMW_IMPLEMENTATION_SUPPORTS_POCO FALSE)
+get_available_rmw_implementations(available_rmw_implementations)
+list(LENGTH available_rmw_implementations count_available_rmw_implementations)
+if(count_available_rmw_implementations EQUAL 1)
+  set(rmw_implementation_baked_default ON)
+else()
+  set(rmw_implementation_baked_default OFF)
 endif()
 
-if(NOT RMW_IMPLEMENTATION_SUPPORTS_POCO)
-  message(STATUS "Dynamic loading disabled; directly referencing '${RMW_IMPLEMENTATION}'")
-  message(STATUS "")
+option(RMW_IMPLEMENTATION_BAKED "\
+  Use only the RMW implementation specified at build time, \
+  ignoring runtime value of the 'RMW_IMPLEMENTATION' environment variable \
+  and the 'RMW_IMPLEMENTATION' CMake option"
+  ${rmw_implementation_baked_default})
 
+if(RMW_IMPLEMENTATION_BAKED)
+  message(STATUS "Dynamic loading disabled; directly referencing '${RMW_IMPLEMENTATION}'")
 else()
-  message(STATUS "Dynamic loading enabled with any of the available rmw implementations")
-  message(STATUS "")
+  # provides FindPoco.cmake and Poco on platforms without it
+  find_package(poco_vendor REQUIRED)
+  find_package(Poco REQUIRED COMPONENTS Foundation)
 
-  find_package(rmw REQUIRED)
+  message(STATUS "Dynamic loading enabled")
 
-  link_directories(${Poco_LIBRARY_DIR})
```

Comment:
Why was this removed?

Reply:
Because it's redundant with ament_target_dependencies(... "Poco" ...)

Reply:
Why would that be redundant? `ament_target_dependencies` doesn't consider `*_LIBRARY_DIR` variables and if `*_LIBRARIES` only contains a name (not an absolute path) this will fail without the link directory.

Reply:
Wait - so then where does it get added to the link_libraries or target_link_libraries? I would expect Poco_LIBRARY_DIR to be set there?

Reply:
`link_directories()` like `include_directories()` changes global state in CMake and is being considered by every target.

Reply:
Okay so where does Poco get added to link_libraries? It should be added to link_directories as close to that point as possible.

Reply:
By the `ament_target_dependencies` call.

Reply:
Which for some reason *doesn’t* also add it to LINK_DIRECTORIES?

Reply:
Using `link_directories` is not recommended imo because it exposes you to the same problem as global include directories: adding one might affect the lookup of any other library. Therefore the ament function doesn't use it. Also only very few packages use that (imo outdated) approach.

Reply:
Okay, so if it's the wrong approach, why are you asking me to use it? I'm so confused :-/. If it's needed, it seems ament_target_dependencies should set it up, as it is part of the dependency. If it's not needed, then neither CMakeLists nor ament_target_dependencies should be setting it.

Also, ${Poco_LIBRARY_DIR} doesn't even seem to be defined when CMake gets to that line, so thhis all may be a moot point.

Reply:
`ament_target_dependencies` doesn't use it since it generates an order problem. That is why this package sets it explicitly using `link_directories` even though this is not great. This package takes the responsibility that it doesn't break the order of searched libraries.

As an alternative you could use the library dir to search for the library name explicitly using `find_library()` and then use the absolute path of the found library.

> ${Poco_LIBRARY_DIR} doesn't even seem to be defined when CMake gets to that line

You would need to verify that on all targeted platforms / supported Poco versions. Also a reference when this has been removed from Poco would be beneficial.

Anyway neither of this is related to the change you are proposing in this PR. Therefore I recommend to not change the way this is currently being done. And if you like to modify it do so in a separate PR where it can be iterated on independently.

Reply:
done. #70